# Feature Selection

The purpose of this notebook is perform feature selection. First we will have to sample the data to 10% of the original observations. This is necessary because the large size of the full feature matrices does not allow for efficient feature selection or for a significant number of random search iterations.

## Roadmap

1. Sample 10% of the training observations randomly
2. Convert numeric columns to `np.float32`
3. Convert boolean columns to `np.uint8`
4. One-hot encode categorical features as necessary
5. Remove one of every pair of columns with all duplicated values (1.0 correlation)
6. Remove columns with more than 90% missing values
7. Remove columns with a single unique value
8. Remove one of every pair of columns with abs(correlation) > 0.95

The reduced data set will then be saved as `_sample.csv` and can be used for 100 iterations of random search with the Gradient Boosting Machine.

In [1]:
import pandas as pd
import numpy as np

# Featuretools Features

In [2]:
feature_matrix = pd.read_csv('../input/feature_matrix.csv', low_memory=False)

# Sampling 10% of the original data
train = feature_matrix[feature_matrix['TARGET'].notnull()].sample(frac = 0.1, random_state = 50)

import gc
gc.enable()
del feature_matrix
gc.collect()

0

### Correct column types

In [3]:
for col in ['SUM(bureau.PREVIOUS_OTHER_LOAN_RATE)', 'SUM(bureau.PREVIOUS_OTHER_LOAN_RATE WHERE CREDIT_ACTIVE = Closed)',
            'SUM(bureau.PREVIOUS_OTHER_LOAN_RATE WHERE CREDIT_ACTIVE = Active)', 'SUM(bureau_balance.bureau.PREVIOUS_OTHER_LOAN_RATE)']:
    try:
        train[col] = train[col].astype(np.float32)
    except:
        print(f'{col} not in data')
    
for col in train:
    if train[col].dtype == 'bool':
        train[col] = train[col].astype(np.uint8)

SUM(bureau.PREVIOUS_OTHER_LOAN_RATE) not in data
SUM(bureau.PREVIOUS_OTHER_LOAN_RATE WHERE CREDIT_ACTIVE = Closed) not in data
SUM(bureau.PREVIOUS_OTHER_LOAN_RATE WHERE CREDIT_ACTIVE = Active) not in data
SUM(bureau_balance.bureau.PREVIOUS_OTHER_LOAN_RATE) not in data


In [4]:
train = pd.get_dummies(train)
n_features_start = train.shape[1] - 2
train.shape

(30751, 2091)

### Columns with duplicated values

In [5]:
x, idx, inv, counts = np.unique(train, axis = 1, return_index = True, return_inverse=True, return_counts=True)
train = train.iloc[:, idx]
n_non_unique_columns = n_features_start - train.shape[1] - 2
train.shape

(30751, 1816)

### Missing Values

The threshold is currently set at 90% but could be lowered. 

In [6]:
missing_threshold = 90

# Find missing and percentage
missing = pd.DataFrame(train.isnull().sum())
missing['percent'] = 100 * (missing[0] / train.shape[0])
missing.sort_values('percent', ascending = False, inplace = True)

# Missing above threshold
missing_cols = list(missing[missing['percent'] > missing_threshold].index)
n_missing_cols = len(missing_cols)

train = train[[x for x in train if x not in missing_cols]]
train.shape

(30751, 1798)

### Zero variance columns

These are any columns with only a single unique value. (`np.nan` does not count as a unique value)

In [7]:
unique_counts = pd.DataFrame(train.nunique()).sort_values(0, ascending = True)
zero_variance_cols = list(unique_counts[unique_counts[0] == 1].index)
n_zero_variance_cols = len(zero_variance_cols)

train = train[[x for x in train if x not in zero_variance_cols]]
train.shape

(30751, 1771)

#### Remove columns containing derivations of target

This is a special consideration we have to take with the Featuretools data because one of the columns is derived from the `TARGET`. (PERCENTILE transformation works on numeric columns of the data and I accidentally left in the `TARGET` when running the feature synthesis).

In [8]:
for col in train:
    if 'TARGET' in col:
        print(col)
        
train.drop(columns = 'PERCENTILE(TARGET)', inplace = True)

TARGET
PERCENTILE(TARGET)


### Find Collinear Variables with Correlation Threshold

The correlation threshold will be set at 0.95 and one out of every pair of columns that are above this threshold will be removed. The column removed is the one that occurs last in the data.

In [9]:
correlation_threshold = 0.95

corr_matrix = train.corr()

# Extract the upper triangle of the correlation matrix
upper = corr_matrix.where(np.triu(np.ones(corr_matrix.shape), k = 1).astype(np.bool))

# Select the features with correlations above the threshold
# Need to use the absolute value
to_drop = [column for column in upper.columns if any(upper[column].abs() > correlation_threshold)]

In [10]:
train = train[[x for x in train if x not in to_drop]]
n_collinear = len(to_drop)
train.shape

(30751, 1042)

In [11]:
n_non_unique_columns

271

In [12]:
n_missing_cols

18

In [13]:
n_zero_variance_cols

27

In [14]:
n_collinear

728

In [15]:
total_removed = n_non_unique_columns + n_missing_cols + n_zero_variance_cols + n_collinear + 1
print('Total columns removed: ', total_removed)

Total columns removed:  1045


In [16]:
train.head()

DAYS_BIRTH  DAYS_REGISTRATION  SUM(installments.DAYS_ENTRY_PAYMENT)  \
83147     -11819.0            -4187.0                                -726.0   
354935    -18794.0            -6010.0                             -157073.0   
64916     -20741.0            -1882.0                              -20122.0   
87122      -9881.0             -932.0                                   NaN   
280351    -14762.0            -6189.0                              -57878.0   

        MIN(previous.SUM(installments.DAYS_ENTRY_PAYMENT))  \
83147                                              -726.0    
354935                                          -122152.0    
64916                                            -13200.0    
87122                                                 NaN    
280351                                           -28635.0    

        MEAN(previous.SUM(installments.DAYS_ENTRY_PAYMENT))  \
83147                                         -726.000000     
354935                                      -39268.250000     
64916                                       -10061.000000     
87122                                                 NaN     
280351                                      -19292.666667     

        MAX(previous.SUM(installments.DAYS_ENTRY_PAYMENT))  DAYS_EMPLOYED  \
83147                                              -726.0          -673.0   
354935                                            -1026.0         -8164.0   
64916                                             -6922.0             NaN   
87122                                                 NaN          -185.0   
280351                                           -10697.0         -2370.0   

        DAYS_ID_PUBLISH  SUM(previous.DAYS_DECISION)  \
83147            -601.0                       -217.0   
354935          -2201.0                      -8634.0   
64916           -4296.0                      -2900.0   
87122           -1076.0                          NaN   
280351          -4187.0                      -7679.0   

        MIN(previous.DAYS_DECISION)           ...             \
83147                        -217.0           ...              
354935                      -2648.0           ...              
64916                       -2041.0           ...              
87122                           NaN           ...              
280351                      -2745.0           ...              

        PERCENTILE(MEAN(bureau.AMT_ANNUITY))  \
83147                                    NaN   
354935                                   NaN   
64916                                    NaN   
87122                                    NaN   
280351                              0.661964   

        MIN(bureau.PERCENTILE(AMT_ANNUITY))  \
83147                                   NaN   
354935                                  NaN   
64916                                   NaN   
87122                                   NaN   
280351                             0.706131   

        MEAN(bureau.PERCENTILE(AMT_ANNUITY))  \
83147                                    NaN   
354935                                   NaN   
64916                                    NaN   
87122                                    NaN   
280351                              0.706131   

        MIN(bureau.NUM_UNIQUE(bureau_balance.STATUS))  \
83147                                             NaN   
354935                                            NaN   
64916                                             NaN   
87122                                             NaN   
280351                                            3.0   

        MEAN(bureau.NUM_UNIQUE(bureau_balance.STATUS))  \
83147                                              NaN   
354935                                             NaN   
64916                                              NaN   
87122                                              NaN   
280351                                             3.0   

        MAX(bureau.NUM_UNIQUE(bureau_balanc

## Save sample of Data

The resulting data has 10% of the training observations and 1042 columns (1040 of which are features). This data is now ready for random search.

In [17]:
train.to_csv('../input/feature_matrix_sample.csv', index = False)

# Function for Feature Selection

We can refactor the four steps completed above into a single function that applies them in the same sequence to any dataframe. This function will be used on the manual and semi-automated feaures with the same inputs as with the Featuretools features.

In [18]:
def feature_selection(feature_matrix, missing_threshold=90, correlation_threshold=0.95):
    """Feature selection for a dataframe."""
    
    feature_matrix = pd.get_dummies(feature_matrix)
    n_features_start = feature_matrix.shape[1]
    print('Original shape: ', feature_matrix.shape)

    _, idx = np.unique(feature_matrix, axis = 1, return_index = True)
    feature_matrix = feature_matrix.iloc[:, idx]
    n_non_unique_columns = n_features_start - feature_matrix.shape[1]
    print('{}  non-unique valued columns.'.format(n_non_unique_columns))

    # Find missing and percentage
    missing = pd.DataFrame(feature_matrix.isnull().sum())
    missing['percent'] = 100 * (missing[0] / feature_matrix.shape[0])
    missing.sort_values('percent', ascending = False, inplace = True)

    # Missing above threshold
    missing_cols = list(missing[missing['percent'] > missing_threshold].index)
    n_missing_cols = len(missing_cols)

    # Remove missing columns
    feature_matrix = feature_matrix[[x for x in feature_matrix if x not in missing_cols]]
    print('{} missing columns with threshold: {}.'.format(n_missing_cols,
                                                                        missing_threshold))
    
    # Zero variance
    unique_counts = pd.DataFrame(feature_matrix.nunique()).sort_values(0, ascending = True)
    zero_variance_cols = list(unique_counts[unique_counts[0] == 1].index)
    n_zero_variance_cols = len(zero_variance_cols)

    # Remove zero variance columns
    feature_matrix = feature_matrix[[x for x in feature_matrix if x not in zero_variance_cols]]
    print('{} zero variance columns.'.format(n_zero_variance_cols))
    
    # Correlations
    corr_matrix = feature_matrix.corr()

    # Extract the upper triangle of the correlation matrix
    upper = corr_matrix.where(np.triu(np.ones(corr_matrix.shape), k = 1).astype(np.bool))

    # Select the features with correlations above the threshold
    # Need to use the absolute value
    to_drop = [column for column in upper.columns if any(upper[column].abs() > correlation_threshold)]

    n_collinear = len(to_drop)
    
    feature_matrix = feature_matrix[[x for x in feature_matrix if x not in to_drop]]
    print('{} collinear columns removed with threshold: {}.'.format(n_collinear,
                                                                          correlation_threshold))
    
    total_removed = n_non_unique_columns + n_missing_cols + n_zero_variance_cols + n_collinear
    
    print('Total columns removed: ', total_removed)
    print('Shape after feature selection: {}.'.format(feature_matrix.shape))
    return feature_matrix

# Manual Features

The process is the same except now we have a function to carry out the feature selection. First we subset to 10% of the training data, and then we apply feature selection. 

In [20]:
manual_features = pd.read_csv('../input/features_manual.csv')
manual_features = manual_features[manual_features['TARGET'].notnull()].sample(frac = 0.1, random_state = 50)

manual_features = feature_selection(manual_features, 90, 0.95)

Original shape:  (30751, 273)
0  non-unique valued columns.
2 missing columns with threshold: 90.
2 zero variance columns.
38 collinear columns removed with threshold: 0.95.
Total columns removed:  42
Shape after feature selection: (30751, 231).


In [21]:
manual_features.head()

DAYS_BIRTH  DAYS_EMPLOYED  PREVIOUS_LOAN_DIFFERENCE_MEAN  \
77158     -14017.0        -3747.0                     -2475.0000   
306191    -16520.0        -4275.0                      3907.5000   
64916     -20741.0            NaN                     -7821.0000   
81133      -9685.0         -318.0                            NaN   
231607    -20891.0         -413.0                     20321.4375   

        DAYS_REGISTRATION  DAYS_ID_PUBLISH  CASH_AVERAGE_LOAN_LENGTH  \
77158             -2384.0          -1046.0                -78.000000   
306191            -3198.0            -82.0                -57.372263   
64916             -1882.0          -4296.0                -51.678571   
81133              -378.0          -1763.0                       NaN   
231607            -3154.0          -3595.0                -38.400000   

        DAYS_LAST_PHONE_CHANGE  FLAG_DOCUMENT_12  FLAG_DOCUMENT_2  \
77158                      0.0               0.0              0.0   
306191                 -1542.0               0.0              0.0   
64916                      0.0               0.0              0.0   
81133                  -1090.0               0.0              0.0   
231607                 -1696.0               0.0              0.0   

        ORGANIZATION_TYPE_Trade: type 5                ...                  \
77158                                 0                ...                   
306191                                0                ...                   
64916                                 0                ...                   
81133                                 0                ...                   
231607                                0                ...                   

        ELEVATORS_MODE  COMMONAREA_MODE  NONLIVINGAPARTMENTS_MODE  \
77158              NaN              NaN                       NaN   
306191             NaN              NaN                       NaN   
64916              NaN              NaN                       NaN   
81133              NaN              NaN                       NaN   
231607             NaN              NaN                       NaN   

        NONLIVINGAPARTMENTS_MEDI  NONLIVINGAREA_MODE  LANDAREA_MODE  \
77158                        NaN                 NaN            NaN   
306191                       NaN                 NaN            NaN   
64916                        NaN                 NaN            NaN   
81133                        NaN                 NaN            NaN   
231607                       NaN                 NaN            NaN   

        LIVINGAPARTMENTS_MODE  FLOORSMIN_AVG  YEARS_BUILD_AVG  \
77158                     NaN            NaN              NaN   
306191                    NaN            NaN              NaN   
64916                     NaN            NaN              NaN   
81133                     NaN            NaN              NaN   
231607                    NaN            NaN              NaN   

        CREDIT_CARD_AMT_BALANCE_MEAN_MEAN  
77158                                 NaN  
306191                                NaN  
64916                                 NaN  
81133                                 NaN  
231607                                NaN  

[5 rows x 231 columns]

Then we can save the features for random search. 

In [22]:
manual_features.to_csv('../input/features_manual_sample.csv', index = False)

# Semi-Automated Features

The final feature matrix is that created by what we called __semi-automated__ feature engineering. The same exact process applies.

In [23]:
semi_features = pd.read_csv('../input/features_semi.csv')
semi_features = semi_features[semi_features['TARGET'].notnull()].sample(frac = 0.1, random_state = 50)

semi_features = feature_selection(semi_features, 90, 0.95)

Original shape:  (30751, 1447)
0  non-unique valued columns.
8 missing columns with threshold: 90.
43 zero variance columns.
516 collinear columns removed with threshold: 0.95.
Total columns removed:  567
Shape after feature selection: (30751, 880).


In [24]:
semi_features.head()

DAYS_BIRTH  IN_CLIENT_IN_LOAN_DAYS_ENTRY_PAYMENT_sum_sum  \
77158     -14017.0                                       -9252.0   
306191    -16520.0                                     -171424.0   
64916     -20741.0                                      -20122.0   
81133      -9685.0                                           NaN   
231607    -20891.0                                      -71851.0   

        IN_CLIENT_IN_LOAN_DAYS_ENTRY_PAYMENT_sum_min  \
77158                                        -9252.0   
306191                                      -39707.0   
64916                                       -13200.0   
81133                                            NaN   
231607                                      -22759.0   

        IN_CLIENT_IN_LOAN_DAYS_ENTRY_PAYMENT_sum_mean  \
77158                                       -9252.000   
306191                                     -17142.400   
64916                                      -10061.000   
81133                                             NaN   
231607                                      -8981.375   

        IN_CLIENT_IN_LOAN_DAYS_ENTRY_PAYMENT_sum_max  DAYS_EMPLOYED  \
77158                                        -9252.0        -3747.0   
306191                                        -638.0        -4275.0   
64916                                        -6922.0            NaN   
81133                                            NaN         -318.0   
231607                                       -1782.0         -413.0   

        BUREAU_DAYS_CREDIT_sum  PREVIOUS_LOAN_DIFFERENCE_MEAN  \
77158                  -2907.0                     -2475.0000   
306191                -29411.0                      3907.5000   
64916                  -4398.0                     -7821.0000   
81133                   -775.0                            NaN   
231607                 -1294.0                     20321.4375   

        DAYS_REGISTRATION  PREVIOUS_DAYS_DECISION_sum  \
77158             -2384.0                     -2375.0   
306191            -3198.0                    -17739.0   
64916             -1882.0                     -2900.0   
81133              -378.0                         NaN   
231607            -3154.0                     -9801.0   

                              ...                          \
77158                         ...                           
306191                        ...                           
64916                         ...                           
81133                         ...                           
231607                        ...                           

        CC_CLIENT_CC_LOAN_AMT_DRAWINGS_ATM_CURRENT_max_sum  \
77158                                                 NaN    
306191                                                NaN    
64916                                                 NaN    
81133                                                 NaN    
231607                                                NaN    

        CC_CLIENT_CC_LOAN_AMT_DRAWINGS_CURRENT_max_min  \
77158                                              NaN   
306191                                             NaN   
64916                                              NaN   
81133                                              NaN   
231607                                             NaN   

        CC_CLIENT_CC_LOAN_AMT_CREDIT_LIMIT_ACTUAL_mean_min  \
77158                                                 NaN    
306191                                                NaN    
64916                                                 NaN    
81133                                                 NaN    
231607                                                NaN    

        CC_CLIENT_CC_LOAN_AMT_PAYMENT_TOTAL_CURRENT_max_min  \
77158                                                 NaN     
306191                                                NaN     
64916                                                 NaN     
81133                                       

In [25]:
semi_features.to_csv('../input/features_semi_sample.csv', index = False)

# Default Features

These are the features available in the main dataframe.

In [26]:
fm = pd.read_csv('../input/application_train.csv')
fm = fm.sample(frac = 0.1, random_state = 50)

fm = pd.get_dummies(fm)
fm.shape

(30751, 244)

In [27]:
fm = feature_selection(fm, 90, 0.95)
fm.head()

Original shape:  (30751, 244)
0  non-unique valued columns.
0 missing columns with threshold: 90.
2 zero variance columns.
37 collinear columns removed with threshold: 0.95.
Total columns removed:  39
Shape after feature selection: (30751, 205).


DAYS_BIRTH  DAYS_EMPLOYED  DAYS_REGISTRATION  DAYS_ID_PUBLISH  \
77158       -14017          -3747            -2384.0            -1046   
306191      -16520          -4275            -3198.0              -82   
64916       -20741         365243            -1882.0            -4296   
81133        -9685           -318             -378.0            -1763   
231607      -20891           -413            -3154.0            -3595   

        DAYS_LAST_PHONE_CHANGE  FLAG_DOCUMENT_12  FLAG_DOCUMENT_2  \
77158                      0.0                 0                0   
306191                 -1542.0                 0                0   
64916                      0.0                 0                0   
81133                  -1090.0                 0                0   
231607                 -1696.0                 0                0   

        ORGANIZATION_TYPE_Trade: type 5  NAME_INCOME_TYPE_Maternity leave  \
77158                                 0                                 0   
306191                                0                                 0   
64916                                 0                                 0   
81133                                 0                                 0   
231607                                0                                 0   

        FLAG_DOCUMENT_7       ...         BASEMENTAREA_MEDI  ELEVATORS_MODE  \
77158                 0       ...                       NaN             NaN   
306191                0       ...                       NaN             NaN   
64916                 0       ...                       NaN             NaN   
81133                 0       ...                       NaN             NaN   
231607                0       ...                       NaN             NaN   

        COMMONAREA_MODE  NONLIVINGAPARTMENTS_MODE  NONLIVINGAPARTMENTS_MEDI  \
77158               NaN                       NaN                       NaN   
306191              NaN                       NaN                       NaN   
64916               NaN                       NaN                       NaN   
81133               NaN                       NaN                       NaN   
231607              NaN                       NaN                       NaN   

        NONLIVINGAREA_MODE  LANDAREA_MODE  LIVINGAPARTMENTS_MODE  \
77158                  NaN            NaN                    NaN   
306191                 NaN            NaN                    NaN   
64916                  NaN            NaN                    NaN   
81133                  NaN            NaN                    NaN   
231607                 NaN            NaN                    NaN   

        FLOORSMIN_AVG  YEARS_BUILD_AVG  
77158             NaN              NaN  
306191            NaN              NaN  
64916             NaN              NaN  
81133             NaN              NaN  
231607            NaN              NaN  

[5 rows x 205 columns]

In [28]:
fm.to_csv('../input/features_default_sample.csv', index = False)

# Conclusions from Feature Selection

In order to allow us to perform feature selection, we first had to limit the number of rows of data to 10% of the training observations. Then we were able to apply feature selection to reduce the data dimensionality. The final sampled and selected data was saved as `_sample.csv` and can now be used for random search with the Gradient Boosting Machine. There are a number of "arbitrary" thresholds in this approach, but by applying the same operations to all 3 datasets, it is hoped that these choices will not affect the integrity of the analysis. At the end of the day, some results are better than no results and these operations will allow us to perform 100 iterations of random search and proceed with the project. 

The next step is Subsetting Data where we use these results in order to create versions of the feature matrices with all the observations but only the columns identified. 

# Subsetting Data

Now that we have identified the features we want to keep, we can subset the full data matrices to these features. The `_sample.csv` files are the sampled version (10% of the observations) with the feature selection applied. The `fm` dataframes are the full data. The full data after being subsetted is then saved as `_selected.csv`.

The random search is run on the `_sample.csv` files, but the `_selected.csv` files will be used to train and test the final model. These files have all the observations, but only the columns that remained after feature selection.

## Baseline Features

In [29]:
train, test = pd.read_csv('../input/application_train.csv'), pd.read_csv('../input/application_test.csv')
test['TARGET'] = np.nan
train, test = pd.get_dummies(train).align(pd.get_dummies(test), axis = 1, join = 'inner')

sample = pd.read_csv('../input/features_default_sample.csv')
fm = train.append(test, ignore_index = True)
fm.shape

(356255, 243)

In [30]:
fm = fm[[x for x in sample.columns if x in fm.columns]]
fm.to_csv('../input/features_default_selected.csv')
fm.shape

(356255, 204)

## Manual Engineered Features

In [31]:
sample = pd.read_csv('../input/features_manual_sample.csv')
fm = pd.read_csv('../input/features_manual.csv')

# One-hot encoding
fm = pd.get_dummies(fm)
fm.shape

(356255, 273)

In [32]:
# Subset to the columns in the sample
fm = fm[sample.columns]
fm.shape

(356255, 231)

## Automated Engineering Features

In [33]:
# Read in sample and full data
sample = pd.read_csv('../input/feature_matrix_sample.csv')
fm = pd.read_csv('../input/feature_matrix.csv')

print(fm.shape)

# One hot encoding
cat = pd.get_dummies(fm.select_dtypes('object'))

# Convert the column types
for col in fm:
    if fm[col].dtype == 'bool':
        fm[col] = fm[col].astype(np.uint8)
        
# Add the one-hot encoded columns
fm = fm.select_dtypes(['number'])
fm = pd.concat([fm, cat], axis = 1)
fm.shape

(356255, 1821)


(356255, 2111)

In [34]:
sample = pd.read_csv('../input/feature_matrix_sample.csv')
sample.shape

(30751, 1042)

In [35]:
# Subset to the columns in the sample
fm = fm[[x for x in sample.columns if x in fm]]
fm.shape

(356255, 1042)

In [36]:
# Save
fm.to_csv('../input/feature_matrix_selected.csv', index = False)

## Semi-Automated Engineered Features

In [37]:
sample = pd.read_csv('../input/features_semi_sample.csv')
fm = pd.read_csv('../input/features_semi.csv')

# One hot encoding
fm = pd.get_dummies(fm)
fm.shape

(356255, 1447)

In [38]:
# Subset to the columns in sample
fm = fm[sample.columns]
fm.shape

(356255, 880)

In [39]:
fm.to_csv('../input/features_semi_selected.csv', index = False)

# Conclusions

The resulting data can now be used for modeling. The columns have been reduced through feature selection but all of the observations remain. The final datasets will be tested both with the default gradient boosting machine hyperparameters and with the optimal hyperparameters found from random search. 

The next step is to run random search on these results. This is implemented in the `random_search.py` script in the scripts directory.